File: lib/services.js
import { estimateFeatures } from '@/lib/engine'

const WEATHER_API_URL = 'https://api.open-meteo.com/v1/forecast'
const ELEVATION_API_URL = 'https://api.open-meteo.com/v1/elevation'
const GEOCODING_API_URL = 'https://geocoding-api.open-meteo.com/v1/search'
const REVERSE_GEOCODING_API_URL = 'https://nominatim.openstreetmap.org/reverse'

function firstValue(values, fallback) {
  return Array.isArray(values) && values.length ? Number(values[0]) : Number(fallback)
}

export async function fetchConditions(latitude, longitude) {
  const fallback = estimateFeatures(latitude, longitude)
  const fallbackContext = {
    temperature_2m: fallback.temperature,
    wind_speed_10m: fallback.wind_speed,
    surface_pressure: fallback.pressure,
    shortwave_radiation_sum: Number((fallback.solar_irradiance * 3.6).toFixed(2)),
    precipitation_sum: Math.max(Number((fallback.water_availability * 0.15).toFixed(2)), 0),
    timezone: 'UTC',
  }

  try {
    const [weatherResponse, elevationResponse] = await Promise.all([
      fetch(
        `${WEATHER_API_URL}?latitude=${latitude}&longitude=${longitude}&current=temperature_2m,surface_pressure,wind_speed_10m&daily=shortwave_radiation_sum,precipitation_sum&timezone=auto&forecast_days=1&wind_speed_unit=ms`,
        { next: { revalidate: 1800 } },
      ),
      fetch(`${ELEVATION_API_URL}?latitude=${latitude}&longitude=${longitude}`, {
        next: { revalidate: 86400 },
      }),
    ])

    if (!weatherResponse.ok || !elevationResponse.ok) {
      throw new Error('External weather service unavailable')
    }

    const weatherPayload = await weatherResponse.json()
    const elevationPayload = await elevationResponse.json()
    const current = weatherPayload.current || {}
    const daily = weatherPayload.daily || {}
    const elevationValues = elevationPayload.elevation || [fallback.elevation]
    const elevation = Number(elevationValues[0] ?? fallback.elevation)
    const shortwaveSumMj = firstValue(daily.shortwave_radiation_sum, fallback.solar_irradiance * 3.6)
    const precipitationSumMm = firstValue(daily.precipitation_sum, Math.max(fallback.water_availability * 0.15, 0))
    const solarIrradiance = Number((shortwaveSumMj / 3.6).toFixed(2))
    const windSpeed = Number(Number(current.wind_speed_10m ?? fallback.wind_speed).toFixed(2))
    const temperature = Number(Number(current.temperature_2m ?? fallback.temperature).toFixed(2))
    const pressure = Number(Number(current.surface_pressure ?? fallback.pressure).toFixed(2))

    const latitudeRadians = latitude * Math.PI / 180
    const longitudeRadians = longitude * Math.PI / 180
    const hydroIndex = Math.min(
      100,
      Math.max(
        5,
        precipitationSumMm * 6 +
          Math.max(elevation, 0) / 120 +
          15 * Math.abs(Math.sin(latitudeRadians * 0.7 - longitudeRadians * 0.25)),
      ),
    )

    return {
      features: {
        solar_irradiance: solarIrradiance,
        wind_speed: windSpeed,
        elevation: Number(elevation.toFixed(2)),
        water_availability: Number(hydroIndex.toFixed(2)),
        temperature,
        pressure,
      },
      fetchedAt: new Date().toISOString(),
      dataSource: 'open-meteo-live',
      weatherContext: {
        temperature_2m: temperature,
        wind_speed_10m: windSpeed,
        surface_pressure: pressure,
        shortwave_radiation_sum: Number(shortwaveSumMj.toFixed(2)),
        precipitation_sum: Number(precipitationSumMm.toFixed(2)),
        timezone: weatherPayload.timezone || 'UTC',
      },
    }
  } catch {
    return {
      features: fallback,
      fetchedAt: new Date().toISOString(),
      dataSource: 'synthetic-fallback',
      weatherContext: fallbackContext,
    }
  }
}

export async function searchLocations(query, count = 5) {
  const trimmed = query.trim()
  if (trimmed.length < 2) {
    return []
  }

  try {
    const response = await fetch(
      `${GEOCODING_API_URL}?name=${encodeURIComponent(trimmed)}&count=${count}&language=en&format=json`,
      { next: { revalidate: 3600 } },
    )
    if (!response.ok) {
      throw new Error('Location search failed')
    }
    const payload = await response.json()
    return (payload.results || []).map((item) => ({
      name: item.name ?? null,
      country: item.country ?? null,
      admin1: item.admin1 ?? null,
      latitude: item.latitude,
      longitude: item.longitude,
      elevation: item.elevation ?? null,
      timezone: item.timezone ?? null,
    }))
  } catch {
    return []
  }
}

export async function reverseLocation(latitude, longitude) {
  const fallbackName = `Selected point (${latitude.toFixed(4)}, ${longitude.toFixed(4)})`

  try {
    const response = await fetch(
      `${REVERSE_GEOCODING_API_URL}?lat=${latitude}&lon=${longitude}&format=jsonv2&zoom=10&addressdetails=1`,
      {
        headers: {
          'User-Agent': 'MapSense/1.0',
        },
        next: { revalidate: 3600 },
      },
    )

    if (!response.ok) {
      throw new Error('Reverse geocoding failed')
    }

    const payload = await response.json()
    const address = payload.address || {}
    const locality =
      address.city ||
      address.town ||
      address.village ||
      address.municipality ||
      address.county ||
      address.state_district
    const admin1 = address.state || address.region || null
    const country = address.country || null
    const name = [locality, admin1, country].filter(Boolean).filter((item, index, array) => array.indexOf(item) === index).join(', ')

    return {
      name: name || fallbackName,
      country,
      admin1,
      latitude,
      longitude,
    }
  } catch {
    return {
      name: fallbackName,
      country: null,
      admin1: null,
      latitude,
      longitude,
    }
  }
}

File: mapsense/services.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from math import radians, sin

import httpx

from .engine import FeatureSample, estimate_features


WEATHER_API_URL = "https://api.open-meteo.com/v1/forecast"
ELEVATION_API_URL = "https://api.open-meteo.com/v1/elevation"
GEOCODING_API_URL = "https://geocoding-api.open-meteo.com/v1/search"
REVERSE_GEOCODING_API_URL = "https://nominatim.openstreetmap.org/reverse"


@dataclass(frozen=True)
class LiveConditions:
    features: FeatureSample
    fetched_at: str
    data_source: str
    weather_context: dict[str, float | str]


def _first_value(values: object, fallback: float) -> float:
    if isinstance(values, list) and values:
        return float(values[0])
    return float(fallback)


class OpenMeteoService:
    def __init__(self, timeout_seconds: float = 12.0) -> None:
        self.timeout_seconds = timeout_seconds

    def fetch_conditions(self, latitude: float, longitude: float) -> LiveConditions:
        fallback = estimate_features(latitude, longitude)
        fallback_context = {
            "temperature_2m": fallback.temperature,
            "wind_speed_10m": fallback.wind_speed,
            "surface_pressure": fallback.pressure,
            "shortwave_radiation_sum": round(fallback.solar_irradiance * 3.6, 2),
            "precipitation_sum": max(round(fallback.water_availability * 0.15, 2), 0.0),
            "timezone": "UTC",
        }

        try:
            with httpx.Client(timeout=self.timeout_seconds) as client:
                weather_response = client.get(
                    WEATHER_API_URL,
                    params={
                        "latitude": latitude,
                        "longitude": longitude,
                        "current": "temperature_2m,surface_pressure,wind_speed_10m",
                        "daily": "shortwave_radiation_sum,precipitation_sum",
                        "timezone": "auto",
                        "forecast_days": 1,
                        "wind_speed_unit": "ms",
                    },
                )
                weather_response.raise_for_status()

                elevation_response = client.get(
                    ELEVATION_API_URL,
                    params={"latitude": latitude, "longitude": longitude},
                )
                elevation_response.raise_for_status()
        except httpx.HTTPError:
            return LiveConditions(
                features=fallback,
                fetched_at=datetime.now(timezone.utc).isoformat(),
                data_source="synthetic-fallback",
                weather_context=fallback_context,
            )

        weather_payload = weather_response.json()
        elevation_payload = elevation_response.json()
        current = weather_payload.get("current", {})
        daily = weather_payload.get("daily", {})
        elevation_values = elevation_payload.get("elevation", [fallback.elevation])
        elevation = float(elevation_values[0]) if elevation_values else fallback.elevation

        shortwave_sum_mj = _first_value(daily.get("shortwave_radiation_sum"), fallback.solar_irradiance * 3.6)
        precipitation_sum_mm = _first_value(daily.get("precipitation_sum"), max(fallback.water_availability * 0.15, 0.0))
        solar_irradiance = round(shortwave_sum_mj / 3.6, 2)
        wind_speed = round(float(current.get("wind_speed_10m", fallback.wind_speed)), 2)
        temperature = round(float(current.get("temperature_2m", fallback.temperature)), 2)
        pressure = round(float(current.get("surface_pressure", fallback.pressure)), 2)

        latitude_radians = radians(latitude)
        longitude_radians = radians(longitude)
        hydro_index = min(
            100.0,
            max(
                5.0,
                precipitation_sum_mm * 6.0
                + max(elevation, 0.0) / 120.0
                + 15.0 * abs(sin(latitude_radians * 0.7 - longitude_radians * 0.25)),
            ),
        )

        return LiveConditions(
            features=FeatureSample(
                solar_irradiance=solar_irradiance,
                wind_speed=wind_speed,
                elevation=round(elevation, 2),
                water_availability=round(hydro_index, 2),
                temperature=temperature,
                pressure=pressure,
            ),
            fetched_at=datetime.now(timezone.utc).isoformat(),
            data_source="open-meteo-live",
            weather_context={
                "temperature_2m": temperature,
                "wind_speed_10m": wind_speed,
                "surface_pressure": pressure,
                "shortwave_radiation_sum": round(shortwave_sum_mj, 2),
                "precipitation_sum": round(precipitation_sum_mm, 2),
                "timezone": weather_payload.get("timezone", "UTC"),
            },
        )

    def search_locations(self, query: str, count: int = 5) -> list[dict[str, object]]:
        if len(query.strip()) < 2:
            return []

        try:
            with httpx.Client(timeout=self.timeout_seconds) as client:
                response = client.get(
                    GEOCODING_API_URL,
                    params={
                        "name": query.strip(),
                        "count": count,
                        "language": "en",
                        "format": "json",
                    },
                )
                response.raise_for_status()
        except httpx.HTTPError:
            return []

        payload = response.json()
        return [
            {
                "name": item.get("name"),
                "country": item.get("country"),
                "admin1": item.get("admin1"),
                "latitude": item.get("latitude"),
                "longitude": item.get("longitude"),
                "elevation": item.get("elevation"),
                "timezone": item.get("timezone"),
            }
            for item in payload.get("results", [])
        ]

    def reverse_location(self, latitude: float, longitude: float) -> dict[str, object]:
        fallback_name = f"Selected point ({round(latitude, 4)}, {round(longitude, 4)})"

        try:
            with httpx.Client(
                timeout=self.timeout_seconds,
                headers={"User-Agent": "MapSense/1.0"},
            ) as client:
                response = client.get(
                    REVERSE_GEOCODING_API_URL,
                    params={
                        "lat": latitude,
                        "lon": longitude,
                        "format": "jsonv2",
                        "zoom": 10,
                        "addressdetails": 1,
                    },
                )
                response.raise_for_status()
        except httpx.HTTPError:
            return {
                "name": fallback_name,
                "country": None,
                "admin1": None,
                "latitude": latitude,
                "longitude": longitude,
            }

        payload = response.json()
        address = payload.get("address", {})
        locality = (
            address.get("city")
            or address.get("town")
            or address.get("village")
            or address.get("municipality")
            or address.get("county")
            or address.get("state_district")
        )
        admin1 = address.get("state") or address.get("region")
        country = address.get("country")
        name_parts = [part for part in [locality, admin1, country] if part]
        name = ", ".join(dict.fromkeys(name_parts)) if name_parts else fallback_name

        return {
            "name": name,
            "country": country,
            "admin1": admin1,
            "latitude": latitude,
            "longitude": longitude,
        }
